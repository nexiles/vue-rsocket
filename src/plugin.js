import Observe from './Observe';
import GlobalEmitter from './GlobalEmitter';
import createMixin from './createMixin';
import { isSocketIo } from './utils';
import defaults from './defaults';

/**
 * @param {Vue} Vue
 * @param {SocketIOClient} socket
 * @param {Object} obj
 * @return {Object}
 */
function defineReactiveProperties(Vue, socket, obj) {
  const configStore = new Vue({
    data: () => ({
      connected: false,
    }),
    computed: {
      disconnected() {
        return !this.connected;
      },
    },
  });
  socket.on('connect', () => {
    configStore.connected = true;
  });
  socket.on('disconnect', () => {
    configStore.connected = false;
  });

  return Object.defineProperties(obj, {
    connected: {
      get() {
        return configStore.connected;
      },
      enumerable: false,
    },
    disconnected: {
      get() {
        return configStore.disconnected;
      },
      enumerable: false,
    },
  });
}

/**
 * @param {SocketIOClient} socket
 * @param {Object} obj
 * @return {Object}
 */
function defineSocketIoClient(socket, obj) {
  return Object.defineProperties(obj, {
    client: {
      value: socket,
      writable: false,
      enumerable: false,
    },
  });
}

// eslint-disable-next-line import/prefer-default-export
function install(Vue, socket, options) {
  if (!isSocketIo(socket)) {
    throw new Error('[vue-socket.io-ext] you have to pass `socket.io-client` instance to the plugin');
  }
  const $socket = {};
  defineReactiveProperties(Vue, socket, $socket);
  defineSocketIoClient(socket, $socket);

  // eslint-disable-next-line no-param-reassign
  Vue.prototype.$socket = $socket;
  Observe(socket, options);
  Vue.mixin(createMixin(GlobalEmitter));
  const strategies = Vue.config.optionMergeStrategies;
  strategies.sockets = strategies.methods;
}

export { defaults, install };
