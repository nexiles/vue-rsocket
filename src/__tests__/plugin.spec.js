import { createLocalVue, mount } from '@vue/test-utils';
import * as Main from '../plugin';
import io from '../__mocks__/socket.io-client';

it('should be vue plugin (is an object with `install` method)', () => {
  expect(Main).toMatchObject({
    install: expect.any(Function),
  });
});

describe('.install()', () => {
  let Vue;
  beforeEach(() => {
    Vue = createLocalVue();
  });

  it('should accept 3 arguments', () => {
    expect(Main.install).toHaveProperty('length', 3);
  });

  it('should throw an error when no second argument passed', () => {
    expect(() => Vue.use(Main))
      .toThrowErrorMatchingSnapshot();
  });

  it('should throw an error when second argument is not socket.io instance', () => {
    expect(() => Vue.use(Main, { a: 1 }))
      .toThrowErrorMatchingSnapshot();
    expect(() => Vue.use(Main, 'ws://localhost'))
      .toThrowErrorMatchingSnapshot();
  });

  it('should not throw an error when second argument is socket.io instance', () => {
    const socket = io('ws://localhost');
    expect(() => Vue.use(Main, socket))
      .not
      .toThrow();
  });

  it('defines socket.io instance as `$socket.client` on Vue prototype', () => {
    Vue.use(Main, io('ws://localhost'));
    const wrapper = mount({ render: () => null }, { localVue: Vue });
    expect(wrapper.vm.$socket.client).toBeDefined();
    expect(wrapper.vm.$socket.client).toEqual(expect.any(Object));
    expect(wrapper.vm.$socket.client.on).toEqual(expect.any(Function));
    expect(wrapper.vm.$socket.client.emit).toEqual(expect.any(Function));
  });

  it('defines reactive property `$socket.connected` on Vue prototype', () => {
    const socket = io('ws://localhost');
    Vue.use(Main, socket);
    const wrapper = mount({ render: () => null }, { localVue: Vue });
    expect(wrapper.vm.$socket.connected).toBeDefined();
    expect(wrapper.vm.$socket.connected).toBe(false);
    socket.fireSystemEvent('connect');
    expect(wrapper.vm.$socket.connected).toBeDefined();
    expect(wrapper.vm.$socket.connected).toBe(true);
    socket.fireSystemEvent('disconnect');
    expect(wrapper.vm.$socket.connected).toBeDefined();
    expect(wrapper.vm.$socket.connected).toBe(false);
  });

  it('defines reactive property `$socket.disconnected` on Vue prototype', () => {
    const socket = io('ws://localhost');
    Vue.use(Main, socket);
    const wrapper = mount({ render: () => null }, { localVue: Vue });
    expect(wrapper.vm.$socket.disconnected).toBeDefined();
    expect(wrapper.vm.$socket.disconnected).toBe(true);
    socket.fireSystemEvent('connect');
    expect(wrapper.vm.$socket.disconnected).toBeDefined();
    expect(wrapper.vm.$socket.disconnected).toBe(false);
    socket.fireSystemEvent('disconnect');
    expect(wrapper.vm.$socket.disconnected).toBeDefined();
    expect(wrapper.vm.$socket.disconnected).toBe(true);
  });

  it('registers mixin on Vue', () => {
    const spy = jest.spyOn(Vue, 'mixin');
    Vue.use(Main, io('ws://localhost'));
    expect(spy).toHaveBeenCalledTimes(1);
    expect(spy).toHaveBeenCalledWith(expect.any(Object));
  });

  it('correctly merges `sockets` property of the mixin and the component', () => {
    Vue.use(Main, io('ws://localhost'));
    const mixinListener = jest.fn();
    const componentListener = jest.fn();
    const someMixin = {
      sockets: { mixinListener },
    };
    const wrapper = mount({
      mixins: [someMixin],
      render: () => null,
      sockets: { componentListener },
    }, { localVue: Vue });

    expect(wrapper.vm.$options.sockets).toMatchObject({
      mixinListener,
      componentListener,
    });
  });
});

describe('.defaults', () => {
  it('is an object', () => {
    expect(Main.defaults).toEqual(expect.any(Object));
  });

  it('contains default plugin options', () => {
    expect(Main.defaults).toMatchObject({
      actionPrefix: expect.any(String),
      mutationPrefix: expect.any(String),
      eventToMutationTransformer: expect.any(Function),
      eventToActionTransformer: expect.any(Function),
    });
  });

  it('all values are read-only', () => {
    Object.keys(Main.defaults)
      .forEach((prop) => {
        const fn = () => {
          Main.defaults[prop] = 'newValue';
        };
        expect(fn).toThrow('Cannot assign to read only property');
      });
  });
});
