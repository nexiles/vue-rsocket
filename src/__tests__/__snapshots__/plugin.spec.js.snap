// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.install() should throw an error when no second argument passed 1`] = `"[vue-socket.io-ext] you have to pass \`socket.io-client\` instance to the plugin"`;

exports[`.install() should throw an error when second argument is not socket.io instance 1`] = `"[vue-socket.io-ext] you have to pass \`socket.io-client\` instance to the plugin"`;

exports[`.install() should throw an error when second argument is not socket.io instance 2`] = `"[vue-socket.io-ext] you have to pass \`socket.io-client\` instance to the plugin"`;
